docs/source/modifying-messages.rst

```diff
@@ -0,0 +1,390 @@
+    enough bytes from a ClientHello with extensions to turn it into one
+    without extensions can cause the
+    :py:func:`~tlsfuzzer.messages.pad_handshake` to create a well-formed
+    message, despite modifying it.
+
+Modifying content
+-----------------
+
+The :py:func:`~tlsfuzzer.messages.fuzz_message` supports changing arbitrary
+parts of sent messages.
+
+Both optional parameters of the function, ``substitutions`` and ``xors`` expect
+a dictionary as a value.
+The keys of the dictionary specify the bytes to change.
+To specify the bytes counting from the end of the message use negative numbers.
+
+For example, to change type of a ClientHello message to that of ServerHello
+use the following code:
+
+.. code:: python
+
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA]
+    exts = {ExtensionType.renegotiation_info: None}
+    msg_gen = ClientHelloGenerator(cipihers, extensions=exts)
+    node = node.add_child(fuzz_message(msg_gen,
+                                       {0: HandshakeType.server_hello}))
+
+You can find example use of it in
+`test-invalid-client-hello.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-invalid-client-hello.py>`_
+script.
+
+Modifying records
+=================
+
+The :term:`TLS` protocol specifies four types of encrypted records:
+ones that use stream encryption, ones that use block encryption in
+:term:`MAC` then encrypt mode, ones that use block encryption in encrypt then
+:term:`MAC` mode, and ones that use :term:`AEAD` ciphers.
+Each of them behaves differently on record layer level, thus modifying the
+intermediate ciphertext requires use of different functions.
+
+Fuzzing the MAC
+---------------
+
+Depending on cipher suite and negotiated extensions, to change the
+authentication tag you need to use different functions.
+
+For ciphers that use :term:`HMAC` you can change the authentication tag using
+the :py:func:`~tlsfuzzer.messages.fuzz_mac` function.
+
+.. note::
+
+    :py:func:`~tlsfuzzer.messages.fuzz_mac` works with stream ciphers and
+    block ciphers in :term:`CBC` mode only. It doesn't work for SSLv2
+    connections though.
+
+You use :py:func:`~tlsfuzzer.messages.fuzz_mac` the same way as you use
+:py:func:`~tlsfuzzer.messages.fuzz_message`: pass message to change as the
+first argument and use the other two to specify bytes to either xor or
+substitute.
+
+For example, to invert the first and last bit of :term:`HMAC` in a record with
+a Finished message use the following code:
+
+.. code:: python
+
+    msg_gen = FinishedGenerator()
+    xors = {0: 0x80, -1: 0x01}
+    node = node.add_child(fuzz_mac(msg_gen, xors=xors))
+
+You can find example use of it in
+`test-fuzzed-MAC.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-fuzzed-MAC.py>`_
+script.
+
+Since both :term:`AEAD` cipher suites and :term:`CBC` cipher suites in encrypt
+then :term:`MAC` mode don't encrypt the authentication tag, you can use the
```

Comment:
I would suggest 'ecnrypt-then-MAC' with dashes or used quotes around it, to make it clear the whole phrase is the name of the mode

Reply:
done